File: examples/python/m2-tool-runtime/bridge.py
"""与 M1 统一协议内容块之间的轻量桥接。"""

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any, Mapping

from tool_types import ToolCall, ToolResult, ToolStatus

# 允许从相邻 M1 示例导入统一协议类型。
_M1 = Path(__file__).resolve().parents[1] / "m1-unified-protocol"
if str(_M1) not in sys.path:
    sys.path.insert(0, str(_M1))

from protocol import (  # noqa: E402
    Message,
    Role,
    ToolDefinition,
    ToolResultBlock,
    ToolUseBlock,
)


def spec_to_tool_definition(spec) -> ToolDefinition:
    """把 Runtime ToolSpec 变成模型可见的 provider-neutral 工具声明。"""

    return ToolDefinition(
        name=spec.name,
        description=spec.description,
        input_schema=dict(spec.input_schema),
        strict=spec.strict,
    )


def tool_use_to_call(block: ToolUseBlock) -> ToolCall:
    """把 M1 ToolUseBlock 转成 Runtime ToolCall。"""

    return ToolCall(call_id=block.id, name=block.name, arguments=dict(block.input))


def call_from_dict(data: Mapping[str, Any]) -> ToolCall:
    """从简单 JSON 对象构造 ToolCall。"""

    call_id = data.get("call_id") or data.get("id")
    name = data.get("name") or data.get("tool")
    arguments = data.get("arguments") or data.get("input") or {}
    if not isinstance(call_id, str) or not call_id.strip():
        raise ValueError("call_id is required")
    if not isinstance(name, str) or not name.strip():
        raise ValueError("name is required")
    if isinstance(arguments, str):
        arguments = json.loads(arguments)
    if not isinstance(arguments, Mapping):
        raise ValueError("arguments must be an object")
    return ToolCall(call_id=call_id, name=name, arguments=dict(arguments))


def result_to_tool_result_block(result: ToolResult) -> ToolResultBlock:
    """把 Runtime 结果编码为 M1 ToolResultBlock。"""

    return ToolResultBlock(
        tool_use_id=result.call_id,
        content=result.as_text(),
        is_error=result.status is ToolStatus.FAILED,
    )


def result_to_message(result: ToolResult) -> Message:
    """构造 role=tool 的统一消息。"""

    return Message(role=Role.TOOL, content=(result_to_tool_result_block(result),))

File: examples/rust/m2-tool-runtime/src/bridge.rs
//! M2 Runtime 与 M1 统一协议之间的显式桥接。

use crate::{ToolCall, ToolError, ToolResult, ToolSpec, ToolStatus};
use m1_unified_protocol::{
    ContentBlock, Message, Role, ToolDefinition, ToolResultBlock, ToolUseBlock,
};

/// 把 Runtime 规格转换成 provider-neutral 工具声明。
pub fn spec_to_tool_definition(spec: &ToolSpec) -> Result<ToolDefinition, ToolError> {
    ToolDefinition::try_new(
        spec.name.clone(),
        spec.description.clone(),
        spec.input_schema.clone(),
        spec.strict,
    )
    .map_err(|error| ToolError::new(error.to_string()))
}

/// 保留 M1 `call_id`，将模型候选动作转换为 Runtime 调用。
pub fn tool_use_to_call(block: &ToolUseBlock) -> ToolCall {
    ToolCall {
        call_id: block.id.clone(),
        name: block.name.clone(),
        arguments: block.input.clone(),
    }
}

/// 把结构化结果编码为下一次模型请求中的工具观察。
pub fn result_to_tool_result_block(result: &ToolResult) -> ToolResultBlock {
    ToolResultBlock {
        tool_use_id: result.call_id.clone(),
        content: result.as_text(),
        is_error: result.status == ToolStatus::Failed,
    }
}

/// 构造 role=tool 的统一消息。
pub fn result_to_message(result: &ToolResult) -> Result<Message, ToolError> {
    Message::try_new(
        Role::Tool,
        vec![ContentBlock::ToolResult(result_to_tool_result_block(
            result,
        ))],
    )
    .map_err(|error| ToolError::new(error.to_string()))
}
